Adds product query support for Category list block

imanish003 opened the pull request:
Adds Product Query support for the atomic category list block.

On the client side, when the Category list Block is used within the Product Query block, the markup will be rendered on the server side - no javascript related to the Category list block will be loaded.

<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->

<!-- Reference any related issues or PRs here -->

Fixes woocommerce/woocommerce-blocks#7329 

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->

### Testing
1. Checkout this branch and run `npm start` 
2. Be sure that you have latest version of the Gutenberg feature plugin installed and activated
3. Create a post/page and add the Product Query block.
4. Add `woocommerce/product-category-list` block to `Post Template` of product query block
    - Verify that product category block works as it was working before.  
6. Now save the post & Open the frontend side and confirm that:
    - No JS related to the Product Category list block is loaded (search in your network tab for `product-category-list-frontend.js`).
    - The UI of the block is identical to on the editor side.
    - That the settings are reflected correctly (eg: custom style).
7. For comparison, switch to trunk and add the All Products block to confirm that the aforementioned product-category-list-frontend.js is loaded.

* [X] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [ ] Feature plugin
* [X] Experimental

imanish003 committed:
Adds product query support for Category list block

On the client side, when the Category list block is used within the
product query bloc, the markup will be rendered on the server side -
No javascript related to Category list block will be rendered.

imanish003 commented:
Hi @gigitux @sunyatasattva 👋 
This is the first time I have written some PHP code. Therefore, I have created a draft PR instead of a normal PR. 
Can you please have a look and provide me with some feedback? 🙂

Also, I noticed that UI is looking a little different between Editor & Frontend:
| Editor | Frontend |
| - | - |
| ![image](https://user-images.githubusercontent.com/16707866/201645757-1622e028-05e3-41ed-859a-7917a25c4034.png) | ![image](https://user-images.githubusercontent.com/16707866/201645812-46e9d6a4-ba4f-4556-a781-25c0faa3a167.png) | 

Can you please help me with this? 🙂

github-actions[bot] committed:
bot: update checkstyle.xml

gigitux commented on src/BlockTypes/ProductCategoryList.php:
```diff
@@ -48,12 +48,57 @@ protected function get_block_type_supports() {
+		foreach ( $product_categories_terms as $product_category_term ) {
+			$output .= '
+				<li class="category-list-item-' . $product_category_term->slug . '">
+					<a href="' . get_term_link( $product_category_term->term_id ) . '">' . $product_category_term->name . '</a>
```
The problem with the spaces has been a little bit tricky. In the end, this is the fix:

```suggestion
					<a href="' . get_term_link( $product_category_term->term_id ) . '">' . $product_category_term->name . '</a></li>
```

I found different links that discuss this problem:
- https://stackoverflow.com/questions/5078239/how-to-remove-the-space-between-inline-inline-block-elements
- https://css-tricks.com/fighting-the-space-between-inline-block-elements/


![image](https://user-images.githubusercontent.com/4463174/201692360-1dcdb19e-2437-41b7-a110-ee095be4afb8.png)

imanish003 replied:
Ah, Got it. Thanks for sharing the link. The info mentioned [here](https://css-tricks.com/fighting-the-space-between-inline-block-elements/#aa-remove-the-spaces) is useful. 🙌🏻

gigitux commented on src/BlockTypes/ProductCategoryList.php:
```diff
@@ -48,12 +48,57 @@ protected function get_block_type_supports() {
+			$output .= '
+				<li class="category-list-item-' . $product_category_term->slug . '">
+					<a href="' . get_term_link( $product_category_term->term_id ) . '">' . $product_category_term->name . '</a>
+				</li>
```
Obviously, if you want to adopt the solution above, you have to remove this tag.

imanish003 replied:
Got it, Updating the highlighted lines to the following fixed it:
```php
			$output .= '
				<li class="category-list-item-' . $product_category_term->slug . '">
					<a href="' . get_term_link( $product_category_term->term_id ) . '">' . $product_category_term->name . '</a></li>
			';

```

gigitux reviewed (changes requested):
Thanks for your work! You did a great job! 💪 

I noticed that the custom style doesn't work on the frontend side.

![image](https://user-images.githubusercontent.com/4463174/201693142-7d7c8f51-bfea-46dc-9a2d-60c21147c122.png)
*editor side*

![image](https://user-images.githubusercontent.com/4463174/201693187-c8806e17-b02b-4089-865b-a861f86c4bb5.png)
*frontend side*

The `StyleAttributesUtils` class helps you generate the right class/CSS to apply to elements.
Also, I noticed we have the same problem when the block is used inside the `All Products` block. Can you check it and open an issue (or fix it)?

The `Categories` word could have a custom translation. Did you check on this block to see if issue #7511 is replicable?

imanish003 committed:
Fix extra space after the category link

I used the solution from following link: https://css-tricks.com/fighting-the-space-between-inline-block-elements/#aa-remove-the-spaces

Here is the content from the above link:

The reason you get the spaces is because, well, you have spaces between the elements (a line break and a few tabs counts as a space, just to be clear). Minimized HTML will solve this problem, or one of these tricks:

```CSS
<ul>
  <li>
   one</li><li>
   two</li><li>
   three</li>
</ul>
```

or

```CSS
<ul>
  <li>one</li
  ><li>two</li
  ><li>three</li>
</ul>
```

or with comments…

```CSS
<ul>
  <li>one</li><!--
  --><li>two</li><!--
  --><li>three</li>
</ul>
```

They’re all pretty funky, but it does the trick.

imanish003 committed:
Fix custom style doesn't work on the frontend side

For more info, check this comment on the PR:
https://github.com/woocommerce/woocommerce-blocks/pull/7675#pullrequestreview-1179267957

In summary, user can set the custom styles like text color, link color, font size, font weight, line height using the editor sidebar.
These styles weren't showing on the frontend side.

imanish003 commented:
Hi @gigitux , Thanks for providing the solution for the spacing issue 🙏 

> Also, I noticed we have the same problem when the block is used inside the All Products block. Can you check it and open an issue (or fix it)?

Looks like this issue is also there on `trunk` branch; therefore, I have created a new issue for this: woocommerce/woocommerce#42526

imanish003 commented:
Hi @gigitux ,

> The Categories word could have a custom translation. Did you check on this block to see if issue https://github.com/woocommerce/woocommerce-blocks/issues/7511 is replicable?

This issue is not replicating for this PR. I believe this might be because I used [following line](https://github.com/woocommerce/woocommerce-blocks/blob/bac8cb2a90e55e78277a1907b3d4189805add186/src/BlockTypes/ProductCategoryList.php#L93):
```
. __( 'Categories:', 'woo-gutenberg-products-block' )
```

Just to be sure that this issue is not replicating, can you please try replicating this issue too?

imanish003 commented:
Hi @gigitux,
I have resolved the issue related to spacing in commit [d3b95f5](https://github.com/woocommerce/woocommerce-blocks/pull/7675/commits/d3b95f51d19c28a6f5265eb12d9f588d200bf386) and related to custom styling in commit [bac8cb2](https://github.com/woocommerce/woocommerce-blocks/pull/7675/commits/bac8cb2a90e55e78277a1907b3d4189805add186). 
Can you please review again? 🙏

imanish003 committed:
Merge branch 'trunk' into add/7329-atomic-category-list-pq-support

imanish003 committed:
Add support for additional CSS class(es)

gigitux commented:
Thanks for your effort, Manish!

Regarding the style, there is a problem with the link color:

![image](https://user-images.githubusercontent.com/4463174/202677741-39a78cb1-e9fd-48d3-bab8-2fe094b5d658.png)

> Looks like this issue is also there on trunk branch; therefore, I have created a new issue for this: https://github.com/woocommerce/woocommerce/issues/42526

Thanks for opening a new issue!


> Just to be sure that this issue is not replicating, can you please try replicating this issue too?

I'm able to replicate this issue. I switched to the German language, and the block isn't translated:

![image](https://user-images.githubusercontent.com/4463174/202683292-655be1d8-5202-460f-aa06-48a2566a0e3d.png)

The translation is available in German language: https://translate.wordpress.org/projects/wp-plugins/woo-gutenberg-products-block/dev/de/default/

imanish003 committed:
Fix "custom style -> link color" not working

To understand the issue in more details please check following comment:
https://github.com/woocommerce/woocommerce-blocks/pull/7675#issuecomment-1319822535

imanish003 commented:
Hi @gigitux, Thanks for the thorough review 🙏🏻

> Regarding the style, there is a problem with the link color:

I have pushed a fix in commit [b046ec5](https://github.com/woocommerce/woocommerce-blocks/pull/7675/commits/b046ec52149bed18c803b43fd4c58ba55f2d0dbb) after our internal discussion in slack: p1669028757623899-slack-C02UBB1EPEF

> I'm able to replicate this issue. I switched to the German language, and the block isn't translated:

As discussed in our huddle call, we already have [following issue](https://github.com/woocommerce/woocommerce-blocks/issues/7511) created in past which we will handle separately.

gigitux reviewed (changes requested):
Thanks for addressing the feedback! 🙇 

There is an error during the save when the `Product Categories List` block doesn't have a custom class.

![image](https://user-images.githubusercontent.com/4463174/203760362-76edd498-d55e-406a-b6a4-236d624e0b95.png)

To avoid this error, we should check on [this line](https://github.com/woocommerce/woocommerce-blocks/pull/7675/files#diff-cba89eae151ce52689cc5094ef24b697acea642796062009a14924d9ca0343c3R91) that `$attributes['className']` is defined.

imanish003 committed:
Fix classname undefined issue

imanish003 commented:
Hi @gigitux,
Nice catch. Thank you 🙏🏻 
I fixed the same issue in [sale badge PR](https://github.com/woocommerce/woocommerce-blocks/pull/7709) too 🙂

I have fixed the issue in commit [f03842e](https://github.com/woocommerce/woocommerce-blocks/pull/7675/commits/f03842e7502346f11b558bee155ca26e92f608e9). Please have a look again 🙏🏻

gigitux reviewed (approved):
Thanks for working on this! Great job! LGTM! 🚢

imanish003 committed:
Merge branch 'trunk' into add/7329-atomic-category-list-pq-support
